### PR TITLE
CCv0 | Multistrap & enable cross-building guest

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
osbuilder: Multistrap Ubuntu

Use `multistrap` for building Ubuntu rootfs. Adds support for building
for foreign architectures using the `ARCH` environment variable
(including umoci).
In the process, the Ubuntu rootfs workflow is vastly simplified.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
packaging: Enable cross-building agent

Requires setting ARCH and CC.

- Add CC linker option for building agent.
- Set host for building libseccomp.

Fixes: #3681
Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
osbuilder: Fix use of LIBC in rootfs.sh

- Add a doc comment
- Pass to build container, e.g. to build x86_64 with glibc (would
  always use musl)

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
osbuilder: Simplify Rust installation

no double export, direct target

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
osbuilder: Remove musl installations

Remove a lot of cruft of musl installations -- we needed those for the
Go agent, but Rustup just takes care of everything. aarch64 on
Debian-based & Alpine is an exception -- create a symlink
`aarch64-linux-musl-gcc` to `musl-tools`'s `musl-gcc` or `gcc` on
Alpine. This is unified -- arch-specific Dockerfiles are removed.

Furthermore, we should keep it in Ubuntu for supporting the offline SEV
KBC.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
osbuilder: apk add --no-cache

Hadolint DL3019. If you're wondering why this is in this PR, that's
because I touch the file later, and we're only triggering the lints for
changed files.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---
Port of #3682. /cc @stevenhorsman

### CODE REVIEW

**Comment by stevenhorsman** [comment]
> @Jakob-Naucke, this was brought in by the weekly rebase. Shall close this one or are there specific bits you want to be part of CCv0?

@fidencio - Our attempt to pull in the multi-strap PR from `main` to `CCv0` as part of the rebase (in [#3932](https://github.com/kata-containers/kata-containers/pull/3932#issuecomment-1076581234)) actually failed as we didn't have Jakob's fixes from this PR in, so after chatting to him we decided to merge in the commit before this from `main`, so this hasn't been included yet. Sorry if I didn't make this clear in the PRs!

**Comment by fidencio** [comment]
@stevenhorsman, hmm. So, this one should have been merged *before* the commit I merged a few minutes ago? :-)

**Comment by stevenhorsman** [comment]
> @stevenhorsman, hmm. So, this one should have been merged _before_ the commit I merged a few minutes ago? :-)

Nope - sorry I've obviously done a bad job again in explaining.
- The merge of `main` into `CCv0` that you merged this morning contained all the commits in `main` up until the multistrap one (so CCv0 is ~ 3 days out of date at the moment). 
- The next step is for Jakob to merge this into `CCv0` will *should* include moving the multistrap commits over to `CCv0`
- Next week's merge of `main` into `CCv0` will then 🤞 pull in everything after the multistrap commit and get us back up to date.

That's my grand plan, but there might be some holes in it!

**Comment by wainersm** [comment]
hi @Jakob-Naucke , I only read the description of this PR so far and one thing caught my attention: the use of an environment called `CC`. Today we export `export CCV0="yes"` to mean "enable CC" stuffs; at some point in time `CCV0` becomes `CC` and, bang, we have a naming clash. :(

I don't know exactly what was the script but I had to unset `VERSION` because it was overwriting the `VERSION` sourced from `/etc/os-release`. Unfortunately Kata doesn't use the idea of "namespaces" on the environment variables exported/used by the scripts.

File: tools/osbuilder/scripts/lib.sh
Comment (wainersm) [inline tools/osbuilder/scripts/lib.sh:304]: Double-checking that even in aarch64 it doesn't need that symlink anymore.

File: ci/install_libseccomp.sh
Comment (wainersm) [inline ci/install_libseccomp.sh:73]: Sorry, didn't get. Why?

File: utils.mk
Comment (wainersm) [inline utils.mk:117]: Nit: document HOST_ARCH move "##VAR ARCH=arch ..." to the line before ARCH.

File: ci/install_libseccomp.sh
Comment (Jakob-Naucke) [reply ci/install_libseccomp.sh:73]: > the use of an environment called `CC`. Today we export `export CCV0="yes"` to mean "enable CC" stuffs; at some point in time `CCV0` becomes `CC` and, bang, we have a naming clash. :(

`CC` (C compiler) is an environment variable used by autotools, so we should avoid it for Confidential Containers. This is also what it is used for here, and this matters e.g. for libseccomp, where we want to build it for the target architecture, and `CC` would be something like `s390x-linux-gnu-gcc`.

> Sorry, didn't get. Why?

Here specifically, we build gperf against the native architecture, as it is a build dependency for libseccomp, and not for the target. Thus, we unset `CC`, which might point to a cross compiler. I can clarify the comment a little.

Comment (jodh-intel) [reply ci/install_libseccomp.sh:73]: Could you summarise that in a comment in the file itself as the existing comment is a bit too terse to derive all that ;)

Also, if this is CC specific, we should state that (which reminds me of the rather unfortunate issue of "CC" being an overloaded acronym in this instance).

File: tools/osbuilder/rootfs-builder/ubuntu/Dockerfile.in
Comment (jodh-intel) [inline tools/osbuilder/rootfs-builder/ubuntu/Dockerfile.in:20]: I want this printed on a T-shirt! :smile:

File: tools/osbuilder/rootfs-builder/ubuntu/rootfs_lib.sh
Comment (jodh-intel) [inline tools/osbuilder/rootfs-builder/ubuntu/rootfs_lib.sh:23]: This makes me slightly twitchy. Paranoia, but might be worth adding some checks at the start of this function:

```bash
[ -z "$rootfs_dir" ] && die "need rootfs"
[ "$rootfs_dir" = "/" ] && die "rootfs cannot be slash"
```
Not _that_ slash!  :tophat: :guitar: :smile:

File: utils.mk
Comment (jodh-intel) [inline utils.mk:149]:
```suggestion
         CC = gcc
         $(warning "WARNING: A foreign ARCH was passed, but no CC alternative. Using $(CC).")
```

**Review by jodh-intel — COMMENTED** [review]
Thanks @Jakob-Naucke. A few comments...

File: ci/install_libseccomp.sh
Comment (Jakob-Naucke) [reply ci/install_libseccomp.sh:73]: This isn't CC specific -- it's a port of #3682. I'll port all additional comments gathered from the review here back into main.

**Comment by fidencio** [comment]
Sorry, I'll use this for a simple test.
/test

**Comment by fidencio** [comment]
@Jakob-Naucke, I am not sure if you're still around, but I must admit I'm confused here.
Parts of this PR are already merged on `main`, while parts of this PR should be specific to the CC effort.

It'd be easier if we can get it in a way that the PRs which are a simple backport come first, and then we have another set doing the changes specific to the CC effort).

/cc @stevenhorsman @BbolroC

**Comment by stevenhorsman** [comment]
I completely agree that most of the patches are similar/identical to each other.
This PR is effectively the merging of the code from those 6 `main` commits plus the changes that 'conflict' with CC & fixes to get it working.
I think I'm still misunderstanding you about how you'd like the split - Do you want the first 5 commits over in one PR and then follow up PR with the last commit after they have merged?
Sorry for being slow!

**Comment by fidencio** [comment]
> I think I'm still misunderstanding you about how you'd like the split - Do you want the first 5 commits over in one PR and then follow up PR with the last commit after they have merged?

So, my bad for not actually trying out on my side and, consequently, not properly explaining what I'd like to see.
I'd like to see the patches that can be clearly backported to be backported and have a reference that those are a backport.  The ones that are not a 1:1 backport, I'd like to see a message explaining what caused the conflicts / changes that didn't allow them to be cleanly backported.

For instance, in the first patch I really would like to see:
```
(cherry picked from commit 5c3e55362442726f9234cd2d1a2382f4fde5dcc4)
```

In the second patch, I'd like to see, in the commit message, something like:
"""
This is not a clean backport of 0072cc2 as the `generate_dockerfile()` function in the CCv0 branch takes into account the INSTALL_AC_KBC parameter, which is not present in the `main` branch.
"""

And so on, and so forth.

If something can be split, we better split, but if we cannot split, we better point out whether the patch is a direct backport and, if not, why it cannot be.

Does this sound reasonable?

**Comment by stevenhorsman** [comment]
> If something can be split, we better split, but if we cannot split, we better point out whether the patch is a direct backport and, if not, why it cannot be.
> 
> Does this sound reasonable?

Maybe I've still missed the point, but I guess my concern is that in the absence of Jakob, who did the change and the `CCv0` updates on them, that this process of re-doing his work and trying to see what works/fails in CCv0 after each commit is going to take a reasonable amount of time and seems to offer little value as the changes are going to a temporary development branch that we hope to kill off soon. Maybe that's not a fair take though?

**Comment by stevenhorsman** [comment]
I guess the other option is that we just merge over all the content from `main` into `CCv0` and in that PR have a separate commit where we try and fix up the bits that are broken based on the changes from this PR?

**Comment by fidencio** [comment]
Well, then go ahead and merge it as it's.

**Review by fidencio — APPROVED** [review]
Approving so we can go ahead with the weekly rebases.

**Pull request merged**
